packages/astro/src/core/cookies/cookies.ts

```diff
@@ -191,6 +191,19 @@ class AstroCookies implements AstroCookiesInterface {
 		}
 	}
 
+	/**
+	 * Merges a new AstroCookies instance into the current instance. Any new cookies
+	 * will be added to the current instance, overwriting any existing cookies with the same name.
+	 */
+	merge(cookies: AstroCookies) {
```

Comment:
I have mixed feelings about this method, I would prefer to **return** a new instance instead of manipulating the existing one. Essentially, try to use a pure/functional approach and avoid side-effects.

I would called this method `fork` instead

Reply:
Why? In this case there are already 2 AstroCookies instances created (once for each response), what does adding a 3rd accomplish? This is not immutable style code, and this function is not mutating the argument, it's mutating itself.

Reply:
I don't think this is the correct fix. I think the fix should be done here:

https://github.com/withastro/astro/blob/bc3c3296bfd44f6ab7cad38dc47a39355a6af4cb/packages/astro/src/core/render-context.ts#L237

And here

https://github.com/withastro/astro/blob/bc3c3296bfd44f6ab7cad38dc47a39355a6af4cb/packages/astro/src/core/render-context.ts#L432

And the code should be 

```js
this.cookies = this.cookies.fork(this.request)
```

`fork` will return a new instance of cookies, and the creation of these new cookies is confined inside the `rewrite` function, while the proposed solution seems to be more broad

Reply:
Ok, that plan makes sense then. 👍

Reply:
I don't think will help with cookies that are set on the response. It only merges the request cookies.

Reply:
I'm happy to change it to a pure function, but I don't think it can be done at the point the cookies are first assigned.

Reply:
What confuses me is the term "rewriting a response" from the initial issue. `Astro.rewrite`, actually, rewrites a `Request`, and you'll get a new `Response`. So I wonder how the issue is related to rewriting? I am a bit confused 😅

Reply:
It refers to responses to a rewritten request. If page `/a` has `Astro.rewrite('/b')`, and `/b` sets a cookie, currently the cookie from b is not returned.